diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/WhiteSpaceRule.kt

```diff
@@ -143,40 +164,74 @@ class WhiteSpaceRule : Rule("horizontal-whitespace") {
+        require(requiredSpacesBefore != null || requiredSpacesAfter != null)
         val spacesBefore = node.treePrev.numWhiteSpaces()
-        val spacesAfter = (node.treeNext ?: node.treeParent.treeNext)  // for `!!` and possibly other postfix expressions treeNext can be null
-                .numWhiteSpaces()
-        val requiredNumSpaces = if (operatorNode.elementType in operatorsWithNoWhitespace) 0 else 1
-        if (spacesBefore != null && spacesBefore != requiredNumSpaces || spacesAfter != null && spacesAfter != requiredNumSpaces) {
-            val freeText = "${node.text} should ${if (requiredNumSpaces == 0) "not " else ""}be surrounded by whitespaces"
+        val spacesAfter = if (requiredSpacesAfter != null) {
+            (node.treeNext
+                    ?: node.treeParent.treeNext)  // for `!!` and possibly other postfix expressions treeNext can be null
+                    .numWhiteSpaces()
+        } else {
+            null
+        }
+        val isErrorBefore = requiredSpacesBefore != null && spacesBefore != null && spacesBefore != requiredSpacesBefore
+        val isErrorAfter = requiredSpacesAfter != null && spacesAfter != null && spacesAfter != requiredSpacesAfter
+        if (isErrorBefore || isErrorAfter) {
+            val freeText = "${node.text} should have" +
+                    getDescription(requiredSpacesBefore != null, requiredSpacesAfter != null, requiredSpacesBefore, requiredSpacesAfter) +
+                    ", but has" +
+                    getDescription(isErrorBefore, isErrorAfter, spacesBefore, spacesAfter)
             WRONG_WHITESPACE.warnAndFix(configRules, emitWarn, isFixMode, freeText, node.startOffset) {
-                node.fixSpaceAround(requiredNumSpaces)
+                node.fixSpaceAround(requiredSpacesBefore, requiredSpacesAfter)
             }
         }
     }
 
-    private fun ASTNode.fixSpaceAround(requiredNumSpaces: Int) {
-        if (requiredNumSpaces == 1) {
-            treePrev.let { if (it.elementType == WHITE_SPACE) it.treePrev else it }.leaveSingleWhiteSpace()
-            leaveSingleWhiteSpace()
-        } else if (requiredNumSpaces == 0) {
-            treePrev.removeIfWhiteSpace()
-            // for `!!` and possibly other postfix expressions treeNext can be null
-            (treeNext ?: treeParent.treeNext).removeIfWhiteSpace()
+    private fun handleEolWhiteSpace(node: ASTNode) {
+        val hasSpaces = node.text.substringBefore('\n').contains(' ')
+        // the second condition corresponds to the last line of file
+        val isEol = node.textContains('\n') || node.psi.parentsWithSelf.all { it.nextSibling == null }
+        if (hasSpaces && isEol) {
+            WRONG_WHITESPACE.warnAndFix(configRules, emitWarn, isFixMode, "there should be no spaces in the end of line", node.startOffset) {
+                (node as LeafElement).replaceWithText(node.text.trimStart(' '))
```

Comment:
I thought you are fixing spaces in the end of line

Reply:
That's because `WHITE_SPACE` node on EOL can have newline in it. Therefore I only trim spaces at the beginning of `WHITE_SPACE` node's text.